File: project-analyzer-java/src/main/java/com/analyzer/module/ApiAnalyzer.java
package com.analyzer.module;

import com.analyzer.model.*;
import com.analyzer.model.AnalysisReport.*;
import com.analyzer.plugin.AnalysisModuleInterface;
import com.analyzer.plugin.LanguagePlugin;

import java.nio.file.Path;
import java.util.*;

public class ApiAnalyzer implements AnalysisModuleInterface {

    private static final Set<String> SWAGGER_ANNOTATIONS = Set.of(
            "api", "apioperation", "apiresponse", "apiresponses",
            "operation", "schema", "tag", "tags", "swagger", "openapi"
    );

    @Override
    public String getName() { return "api"; }

    @Override
    public Object analyze(ProjectProfile profile, List<LanguagePlugin> plugins) {
        List<ApiEndpoint> rawEndpoints = collectEndpoints(profile, plugins);
        enrichWithSwagger(rawEndpoints, profile, plugins);
        List<ApiEndpoint> endpoints = deduplicateEndpoints(rawEndpoints);
        List<ApiGroup> groups = groupEndpoints(endpoints);

        ApiResult result = new ApiResult();
        result.setEndpoints(endpoints);
        result.setGroups(groups);
        result.setTotalCount(endpoints.size());
        return result;
    }

    private List<ApiEndpoint> collectEndpoints(ProjectProfile profile, List<LanguagePlugin> plugins) {
        List<ApiEndpoint> all = new ArrayList<>();
        Set<String> processed = new HashSet<>();
        for (LanguagePlugin plugin : plugins) {
            for (ModuleInfo mod : plugin.identifyModules(Path.of(profile.getProjectPath()))) {
                for (String filePath : mod.getKeyFiles()) {
                    if (processed.add(filePath)) {
                        all.addAll(plugin.identifyApis(Path.of(filePath)));
                    }
                }
            }
        }
        return all;
    }

    private void enrichWithSwagger(List<ApiEndpoint> endpoints, ProjectProfile profile,
                                   List<LanguagePlugin> plugins) {
        Map<String, List<ApiEndpoint>> epMap = new HashMap<>();
        for (ApiEndpoint ep : endpoints) {
            String key = (ep.getHandlerClass() + "." + ep.getHandlerMethod()).toLowerCase();
            epMap.computeIfAbsent(key, k -> new ArrayList<>()).add(ep);
        }
        if (epMap.isEmpty()) return;

        Set<String> parsed = new HashSet<>();
        for (LanguagePlugin plugin : plugins) {
            for (ModuleInfo mod : plugin.identifyModules(Path.of(profile.getProjectPath()))) {
                for (String filePath : mod.getKeyFiles()) {
                    if (parsed.add(filePath)) {
                        for (AstNode node : plugin.parseFile(Path.of(filePath))) {
                            extractSwaggerFromNode(node, "", epMap);
                        }
                    }
                }
            }
        }
    }

    private void extractSwaggerFromNode(AstNode node, String parentClass,
                                        Map<String, List<ApiEndpoint>> epMap) {
        String currentClass = (node.getType() == AstNode.AstNodeType.CLASS
                || node.getType() == AstNode.AstNodeType.INTERFACE) ? node.getName() : parentClass;

        if (node.getType() == AstNode.AstNodeType.METHOD || node.getType() == AstNode.AstNodeType.FUNCTION) {
            String key = (currentClass + "." + node.getName()).toLowerCase();
            List<ApiEndpoint> matching = epMap.get(key);
            if (matching != null && node.getAnnotations() != null) {
                for (AstNode.Annotation ann : node.getAnnotations()) {
                    if (SWAGGER_ANNOTATIONS.contains(ann.getName().toLowerCase()) && ann.getAttributes() != null) {
                        String desc = ann.getAttributes().getOrDefault("value",
                                ann.getAttributes().getOrDefault("summary",
                                        ann.getAttributes().getOrDefault("description", null)));
                        if (desc != null) {
                            for (ApiEndpoint ep : matching) {
                                if (ep.getDescription() == null) ep.setDescription(desc);
                            }
                        }
                    }
                }
            }
        }
        if (node.getChildren() != null) {
            for (AstNode child : node.getChildren()) {
                extractSwaggerFromNode(child, currentClass, epMap);
            }
        }
    }

    private List<ApiEndpoint> deduplicateEndpoints(List<ApiEndpoint> endpoints) {
        Map<String, ApiEndpoint> seen = new LinkedHashMap<>();
        for (ApiEndpoint ep : endpoints) {
            String key = ep.getMethod() + ":" + ep.getPath();
            if (!seen.containsKey(key)) {
                seen.put(key, ep);
            } else {
                ApiEndpoint existing = seen.get(key);
                if (existing.getDescription() == null && ep.getDescription() != null) {
                    existing.setDescription(ep.getDescription());
                }
            }
        }
        return new ArrayList<>(seen.values());
    }

    private List<ApiGroup> groupEndpoints(List<ApiEndpoint> endpoints) {
        Map<String, List<ApiEndpoint>> groupMap = new LinkedHashMap<>();
        for (ApiEndpoint ep : endpoints) {
            String name = ep.getHandlerClass() != null && !ep.getHandlerClass().isEmpty()
                    ? ep.getHandlerClass() : "default";
            groupMap.computeIfAbsent(name, k -> new ArrayList<>()).add(ep);
        }
        List<ApiGroup> groups = new ArrayList<>();
        for (var entry : groupMap.entrySet()) {
            ApiGroup g = new ApiGroup();
            g.setName(entry.getKey());
            g.setBasePath(inferBasePath(entry.getValue()));
            g.setEndpoints(entry.getValue());
            groups.add(g);
        }
        return groups;
    }

    private String inferBasePath(List<ApiEndpoint> endpoints) {
        if (endpoints.isEmpty()) return null;
        if (endpoints.size() == 1) {
            String[] parts = endpoints.get(0).getPath().split("/");
            List<String> filtered = Arrays.stream(parts).filter(s -> !s.isEmpty()).toList();
            if (filtered.size() > 1) return "/" + String.join("/", filtered.subList(0, filtered.size() - 1));
            return endpoints.get(0).getPath();
        }
        String[][] segments = endpoints.stream()
                .map(ep -> Arrays.stream(ep.getPath().split("/")).filter(s -> !s.isEmpty()).toArray(String[]::new))
                .toArray(String[][]::new);
        int minLen = Arrays.stream(segments).mapToInt(s -> s.length).min().orElse(0);
        List<String> common = new ArrayList<>();
        for (int i = 0; i < minLen; i++) {
            String seg = segments[0][i];
            boolean allMatch = true;
            for (String[] s : segments) {
                if (!s[i].equals(seg)) { allMatch = false; break; }
            }
            if (allMatch) common.add(seg); else break;
        }
        return common.isEmpty() ? null : "/" + String.join("/", common);
    }
}

File: project-analyzer-ts/src/modules/api-analyzer.ts
/**
 * ApiAnalyzer — 接口路径分析模块
 *
 * 从语言插件收集 HTTP API 接口定义，去重、分组，
 * 提取 Swagger/OpenAPI 描述信息。
 */

import type { AnalysisModuleInterface, ModuleResult } from '../plugins/analysis-module.js';
import type { LanguagePlugin } from '../plugins/language-plugin.js';
import type {
  ProjectProfile,
  ApiResult,
  ApiGroup,
  ApiEndpoint,
  AstNode,
} from '../models/index.js';

/** Swagger/OpenAPI annotation names (case-insensitive lookup). */
const SWAGGER_ANNOTATIONS = new Set([
  'api', 'apioperation', 'apiresponse', 'apiresponses',
  'operation', 'schema', 'tag', 'tags',
  'swagger', 'openapi',
]);

export class ApiAnalyzer implements AnalysisModuleInterface {
  getName(): string {
    return 'api';
  }

  async analyze(profile: ProjectProfile, plugins: LanguagePlugin[]): Promise<ModuleResult> {
    // 1. Collect endpoints from all plugins for every source file
    const rawEndpoints = this.collectEndpoints(profile, plugins);

    // 2. Enrich endpoints with Swagger/OpenAPI descriptions from AST
    this.enrichWithSwaggerDescriptions(rawEndpoints, profile, plugins);

    // 3. Deduplicate by path + method
    const endpoints = this.deduplicateEndpoints(rawEndpoints);

    // 4. Group by handler class (Controller / module name)
    const groups = this.groupEndpoints(endpoints);

    const result: ApiResult = {
      endpoints,
      groups,
      totalCount: endpoints.length,
    };

    return result;
  }

  /**
   * Collect API endpoints from all plugins by calling identifyApis on each source file.
   */
  private collectEndpoints(profile: ProjectProfile, plugins: LanguagePlugin[]): ApiEndpoint[] {
    const allEndpoints: ApiEndpoint[] = [];
    const processedFiles = new Set<string>();

    for (const plugin of plugins) {
      const modules = plugin.identifyModules(profile.projectPath);
      for (const mod of modules) {
        for (const filePath of mod.keyFiles) {
          if (processedFiles.has(filePath)) continue;
          processedFiles.add(filePath);

          const endpoints = plugin.identifyApis(filePath);
          allEndpoints.push(...endpoints);
        }
      }
    }

    return allEndpoints;
  }

  /**
   * Enrich endpoints with Swagger/OpenAPI description from AST annotations.
   * Looks for @ApiOperation, @Operation, etc. on methods matching the handler.
   */
  private enrichWithSwaggerDescriptions(
    endpoints: ApiEndpoint[],
    profile: ProjectProfile,
    plugins: LanguagePlugin[],
  ): void {
    // Build a map of handlerClass.handlerMethod → endpoint for quick lookup
    const endpointMap = new Map<string, ApiEndpoint[]>();
    for (const ep of endpoints) {
      const key = `${ep.handlerClass}.${ep.handlerMethod}`.toLowerCase();
      const list = endpointMap.get(key) ?? [];
      list.push(ep);
      endpointMap.set(key, list);
    }

    if (endpointMap.size === 0) return;

    const parsedFiles = new Set<string>();

    for (const plugin of plugins) {
      const modules = plugin.identifyModules(profile.projectPath);
      for (const mod of modules) {
        for (const filePath of mod.keyFiles) {
          if (parsedFiles.has(filePath)) continue;
          parsedFiles.add(filePath);

          const nodes = plugin.parseFile(filePath);
          for (const node of nodes) {
            this.extractSwaggerFromNode(node, '', endpointMap);
          }
        }
      }
    }
  }

  /**
   * Recursively walk AST nodes looking for Swagger/OpenAPI annotations.
   */
  private extractSwaggerFromNode(
    node: AstNode,
    parentClassName: string,
    endpointMap: Map<string, ApiEndpoint[]>,
  ): void {
    const currentClass = (node.type === 'class' || node.type === 'interface')
      ? node.name
      : parentClassName;

    if (node.type === 'method' || node.type === 'function') {
      const key = `${currentClass}.${node.name}`.toLowerCase();
      const matchingEndpoints = endpointMap.get(key);

      if (matchingEndpoints) {
        for (const ann of node.annotations) {
          if (SWAGGER_ANNOTATIONS.has(ann.name.toLowerCase())) {
            const description = ann.attributes['value']
              ?? ann.attributes['summary']
              ?? ann.attributes['description'];
            if (description) {
              for (const ep of matchingEndpoints) {
                if (!ep.description) {
                  ep.description = description;
                }
              }
            }

            // Extract tags
            const tagValue = ann.attributes['tags'] ?? ann.attributes['tag'];
            if (tagValue) {
              const tags = tagValue.split(',').map((t) => t.trim()).filter(Boolean);
              for (const ep of matchingEndpoints) {
                for (const tag of tags) {
                  if (!ep.tags.includes(tag)) {
                    ep.tags.push(tag);
                  }
                }
              }
            }
          }
        }
      }
    }

    for (const child of node.children) {
      this.extractSwaggerFromNode(child, currentClass, endpointMap);
    }
  }

  /**
   * Deduplicate endpoints by path + method combination.
   * Keeps the first occurrence (which has the most complete info).
   */
  private deduplicateEndpoints(endpoints: ApiEndpoint[]): ApiEndpoint[] {
    const seen = new Map<string, ApiEndpoint>();

    for (const ep of endpoints) {
      const key = `${ep.method}:${ep.path}`;
      if (!seen.has(key)) {
        seen.set(key, ep);
      } else {
        // Merge description/tags from duplicate if the first one is missing them
        const existing = seen.get(key)!;
        if (!existing.description && ep.description) {
          existing.description = ep.description;
        }
        for (const tag of ep.tags) {
          if (!existing.tags.includes(tag)) {
            existing.tags.push(tag);
          }
        }
      }
    }

    return Array.from(seen.values());
  }

  /**
   * Group endpoints by handlerClass (Controller / module name).
   */
  private groupEndpoints(endpoints: ApiEndpoint[]): ApiGroup[] {
    const groupMap = new Map<string, ApiEndpoint[]>();

    for (const ep of endpoints) {
      const groupName = ep.handlerClass || 'default';
      const list = groupMap.get(groupName) ?? [];
      list.push(ep);
      groupMap.set(groupName, list);
    }

    const groups: ApiGroup[] = [];
    for (const [name, eps] of groupMap) {
      const basePath = this.inferBasePath(eps);
      groups.push({
        name,
        basePath,
        endpoints: eps,
      });
    }

    return groups;
  }

  /**
   * Infer a common base path from a group of endpoints.
   */
  private inferBasePath(endpoints: ApiEndpoint[]): string | undefined {
    if (endpoints.length === 0) return undefined;

    const paths = endpoints.map((ep) => ep.path);
    if (paths.length === 1) {
      // For a single endpoint, use the path up to the last segment
      const parts = paths[0].split('/').filter(Boolean);
      if (parts.length > 1) {
        return '/' + parts.slice(0, -1).join('/');
      }
      return paths[0];
    }

    // Find common prefix among all paths
    const segments = paths.map((p) => p.split('/').filter(Boolean));
    const minLen = Math.min(...segments.map((s) => s.length));
    const commonParts: string[] = [];

    for (let i = 0; i < minLen; i++) {
      const seg = segments[0][i];
      if (segments.every((s) => s[i] === seg)) {
        commonParts.push(seg);
      } else {
        break;
      }
    }

    return commonParts.length > 0 ? '/' + commonParts.join('/') : undefined;
  }
}
